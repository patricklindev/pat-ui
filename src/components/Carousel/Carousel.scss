.Carousel {
  position: relative;
  box-shadow: inset 0px 0px 20px 10px rgba(0, 0, 0, 0.2);
  height: 300px;
  width: 500px;
}

.prev,
.next {
  cursor: pointer;
  position: absolute;
  top: 50%;
  width: auto;
  padding: 16px;
  margin-top: -22px;
  color: white;
  font-weight: bold;
  font-size: 18px;
  transition: 0.6s ease;
  border-radius: 0 3px 3px 0;
  user-select: none;
}

.next {
  right: 0;
  border-radius: 3px 0 0 3px;
}

.dots {
  position: absolute;
  bottom: 0;
  left: 50%;
  transform: translate(-50%);
}
.dot {
  cursor: pointer;
  height: 15px;
  width: 15px;
  margin: 0 2px;
  background-color: #bbb;
  border-radius: 50%;
  display: inline-block;
  transition: background-color 0.6s ease;
}

.active,
.dot:hover {
  background-color: #717171;
}
