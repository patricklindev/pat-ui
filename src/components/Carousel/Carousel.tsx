import React, { useState } from 'react';
import { JsxElement } from 'typescript';

interface ICarousel {
  children: JSX.Element | JSX.Element[];

  style?: object;
}

const Carousel: React.FunctionComponent<ICarousel> = (props: ICarousel) => {
  const [counter, setCounter] = useState(0);

  const { children } = props;
  const [length, setLength] = useState(
    Array.isArray(children) ? children.length : 1
  );
  let classNameList: string[] = [];
  console.log('c', children);

  const classNames = classNameList.join('');

  const handleClick = (event: React.MouseEvent) => {
    //console.log(event.currentTarget.id)
    setCounter(parseInt(event.currentTarget.id));
  };
  const handleClickPrev = () => {
    if (counter > 1) {
      setCounter(counter - 1);
    } else {
      setCounter(length - 1);
    }
  };
  const handleClickNext = () => {
    if (counter < length - 1) {
      setCounter(counter + 1);
    } else {
      setCounter(0);
    }
  };

  //deal with children

  return children && Array.isArray(children) ? (
    <div
      className="Carousel"
      style={{
        ...props.style,
        backgroundImage: `url(${children[counter].props.children.props.src})`,
        backgroundSize: `${children[counter].props.children.props.style.width}`,
        backgroundRepeat: 'no-repeat',
      }}
    >
      <a className="prev" onClick={handleClickPrev}>
        &#10094;
      </a>
      <a className="next" onClick={handleClickNext}>
        &#10095;
      </a>
      <div className="center dots">
        {children.map((src, index) => (
          <span
            id={index.toString()}
            className={index === counter ? 'dot active' : 'dot'}
            onClick={handleClick}
          ></span>
        ))}
      </div>
    </div>
  ) : null;
};
export default Carousel;
