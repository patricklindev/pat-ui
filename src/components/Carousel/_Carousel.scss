$my-max-width: 600px;

.Carousel {
  position: relative;
  box-shadow: inset 0px 0px 20px 10px rgba(0, 0, 0, 0.2);
  overflow: hidden;
}

.Carousel__img-wrapper {
  position: relative;

  width: 3 * $my-max-width;
}
.Carousel__img-wrapper img {
  max-width: $my-max-width;
}

#Carousel__img__left {
  position: absolute;
  top: 0;
  left: -$my-max-width;
}
#Carousel__img__center {
  position: absolute;
  top: 0;
  left: 0;
}
#Carousel__img__right {
  position: absolute;
  top: 0;
  left: $my-max-width;
}

/*deal with the situation when prev is clicked*/
#Carousel__img__left-to-center {
  position: absolute;
  display: block;
  animation-name: moveLefttoCenter;
  animation-duration: 0.6s;
  top: 0; /*end with this position after animation*/
  left: 0; /*end with this position after animation*/
}
#Carousel__img__center-to-right {
  position: absolute;
  display: block;
  animation-name: moveCentertoRight;
  animation-duration: 0.6s;
  top: 0; /*end with this position after animation*/
  left: $my-max-width; /*end with this position after animation*/
}
#Carousel__img__mostleft-to-left {
  position: absolute;
  display: block;
  animation-name: moveMostLefttoLeft;
  animation-duration: 0.6s;
  top: 0; /*end with this position after animation*/
  left: -$my-max-width; /*end with this position after animation*/
}
/*deal with with the situation when next is clicked*/
#Carousel__img__right-to-center {
  position: absolute;
  display: block;
  animation-name: moveRighttoCenter;
  animation-duration: 0.6s;
  top: 0; /*end with this position after animation*/
  left: 0; /*end with this position after animation*/
}
#Carousel__img__center-to-left {
  position: absolute;
  display: block;
  animation-name: moveCentertoLeft;
  animation-duration: 0.6s;
  top: 0; /*end with this position after animation*/
  left: -$my-max-width; /*end with this position after animation*/
}
#Carousel__img__mostright-to-right {
  position: absolute;
  display: block;
  animation-name: moveMostRighttoRight;
  animation-duration: 0.6s;
  top: 0; /*end with this position after animation*/
  left: $my-max-width; /*end with this position after animation*/
}
@keyframes moveLefttoCenter {
  0% {
    left: -$my-max-width;
    top: 0px;
  }
  100% {
    left: 0px;
    top: 0px;
  }
}

@keyframes moveCentertoRight {
  0% {
    top: 0;
    left: 0px;
  }
  100% {
    top: 0px;
    left: $my-max-width;
  }
}

@keyframes moveRighttoCenter {
  0% {
    left: $my-max-width;
    top: 0px;
  }
  100% {
    left: 0px;
    top: 0px;
  }
}

@keyframes moveCentertoLeft {
  0% {
    top: 0;
    left: 0px;
  }
  100% {
    top: 0px;
    left: -$my-max-width;
  }
}

@keyframes moveMostLeftToLeft {
  0% {
    left: -2 * $my-max-width;
    top: 0px;
  }
  100% {
    left: -$my-max-width;
    top: 0px;
  }
}

@keyframes moveMostRighttoRight {
  0% {
    left: 2 * $my-max-width;
    top: 0px;
  }
  100% {
    left: $my-max-width;
    top: 0px;
  }
}

.Carousel__prev,
.Carousel__next {
  cursor: pointer;
  position: absolute;
  top: 50%;
  width: auto;
  padding: 16px;
  margin-top: -22px;
  color: white;
  font-weight: bold;
  font-size: 18px;
  border-radius: 0 3px 3px 0;
  user-select: none;
}

.Carousel__prev {
  left: 0;
  border-radius: 3px 0 0 3px;
}

.Carousel__next {
  right: 0;
  border-radius: 3px 0 0 3px;
}

.Carousel__dots {
  position: absolute;
  bottom: 0;
  left: 50%;
  transform: translate(-50%);
}
.Carousel__dot {
  cursor: pointer;
  height: 10px;
  width: 10px;
  margin: 0 2px;
  background-color: #bbb;
  border-radius: 50%;
  display: inline-block;
}

.Carousel__dot-active,
.Carousel__dot:hover {
  background-color: white;
}
