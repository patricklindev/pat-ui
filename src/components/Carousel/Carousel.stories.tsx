import React from 'react';
import { action } from '@storybook/addon-actions';
import Carousel from './Carousel';

export default {
  title: 'Carousel',
  component: Carousel,
};

const imageSrc4 = [
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/cd7a1aaea8e1c5e3d26fe2591e561798.png~tplv-uwbnlip3yd-webp.webp',
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/6480dbc69be1b5de95010289787d64f1.png~tplv-uwbnlip3yd-webp.webp',
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/0265a04fddbd77a19602a15d9d55d797.png~tplv-uwbnlip3yd-webp.webp',
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/24e0dd27418d2291b65db1b21aa62254.png~tplv-uwbnlip3yd-webp.webp',
];

const imageSrc2 = [
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/cd7a1aaea8e1c5e3d26fe2591e561798.png~tplv-uwbnlip3yd-webp.webp',
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/6480dbc69be1b5de95010289787d64f1.png~tplv-uwbnlip3yd-webp.webp',
];

const imageSrc1 = [
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/cd7a1aaea8e1c5e3d26fe2591e561798.png~tplv-uwbnlip3yd-webp.webp',
];

const buttonStyle: React.CSSProperties = {
  marginRight: '5px',
  marginTop: '5px',
};

export const DefaultCarousel = () => (
  <div>
    <p>More than 2 images</p>
    <Carousel
      autoPlay={false}
      style={{
        width: 600,
        height: 240,
      }}
    >
      {imageSrc4.map((src, index) => (
        <div key={index}>
          <img src={src} style={{ width: '100%' }} />
        </div>
      ))}
    </Carousel>

    <br />

    <p>2 images</p>
    <Carousel
      autoPlay={false}
      style={{
        width: 600,
        height: 240,
      }}
    >
      {imageSrc2.map((src, index) => (
        <div key={index}>
          <img src={src} style={{ width: '100%' }} />
        </div>
      ))}
    </Carousel>

    <br />

    <p>1 image</p>
    <Carousel
      autoPlay={false}
      style={{
        width: 600,
        height: 240,
      }}
    >
      {imageSrc1.map((src, index) => (
        <div key={index}>
          <img src={src} style={{ width: '100%' }} />
        </div>
      ))}
    </Carousel>
  </div>
);

export const AutoPlayCarousel = () => (
  <div>
    <p>
      When mouse moves into the Carousel component, auto play is paused. When
      mouse moves out of Carousel component again, auto play resumes.{' '}
    </p>
    <Carousel
      autoPlay={true}
      style={{
        width: 600,
        height: 240,
      }}
    >
      {imageSrc4.map((src, index) => (
        <div key={index}>
          <img src={src} style={{ width: '100%' }} />
        </div>
      ))}
    </Carousel>
  </div>
);

export const DiffSizeCarousel = () => (
  <div>
    <Carousel
      autoPlay={true}
      style={{
        width: 400,
        height: 200,
      }}
    >
      {imageSrc4.map((src, index) => (
        <div key={index}>
          <img src={src} style={{ width: '100%' }} />
        </div>
      ))}
    </Carousel>

    <br />

    <Carousel
      autoPlay={true}
      style={{
        width: 600,
        height: 240,
      }}
    >
      {imageSrc4.map((src, index) => (
        <div key={index}>
          <img src={src} style={{ width: '100%' }} />
        </div>
      ))}
    </Carousel>
  </div>
);
