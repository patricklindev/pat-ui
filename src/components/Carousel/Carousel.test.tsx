import React from 'react';

import { render, fireEvent, screen } from '@testing-library/react';
import Carousel from './Carousel';

const imageSrc4 = [
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/cd7a1aaea8e1c5e3d26fe2591e561798.png~tplv-uwbnlip3yd-webp.webp',
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/6480dbc69be1b5de95010289787d64f1.png~tplv-uwbnlip3yd-webp.webp',
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/0265a04fddbd77a19602a15d9d55d797.png~tplv-uwbnlip3yd-webp.webp',
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/24e0dd27418d2291b65db1b21aa62254.png~tplv-uwbnlip3yd-webp.webp',
];

const imageSrc2 = [
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/cd7a1aaea8e1c5e3d26fe2591e561798.png~tplv-uwbnlip3yd-webp.webp',
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/6480dbc69be1b5de95010289787d64f1.png~tplv-uwbnlip3yd-webp.webp',
];

const imageSrc1 = [
  '//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/cd7a1aaea8e1c5e3d26fe2591e561798.png~tplv-uwbnlip3yd-webp.webp',
];

describe('Carousel', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Carousel
        style={{
          width: 600,
          height: 240,
        }}
      >
        {imageSrc4.map((src, index) => (
          <div key={index}>
            <img
              src={src}
              style={{
                width: '100%',
              }}
            />
          </div>
        ))}
      </Carousel>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('If user passes Carousel component N inner elements following the correct syntax as below. If N>=3, it should render a carousel that has 3 images and N dots', () => {
    const wrapper = render(
      <Carousel
        style={{
          width: 600,
          height: 240,
        }}
      >
        {imageSrc4.map((src, index) => (
          <div key={index}>
            <img
              src={src}
              style={{
                width: '100%',
              }}
            />
          </div>
        ))}
      </Carousel>
    );

    const container = document.querySelector('.Carousel');
    expect(container).toBeInTheDocument();

    const dots = document.querySelector('.Carousel__dots');
    expect(dots).toBeInTheDocument();
    expect(dots?.children.length).toEqual(4);

    const imgs = document.querySelector('.Carousel__img-wrapper');
    expect(imgs).toBeInTheDocument();
    expect(imgs?.children.length).toEqual(3);
  });

  it('If user passes Carousel component N inner elements following the correct syntax as below. If N=2, it should render a carousel that has 2 images and N dots', () => {
    const wrapper = render(
      <Carousel
        style={{
          width: 600,
          height: 240,
        }}
      >
        {imageSrc2.map((src, index) => (
          <div key={index}>
            <img
              src={src}
              style={{
                width: '100%',
              }}
            />
          </div>
        ))}
      </Carousel>
    );

    const container = document.querySelector('.Carousel');
    expect(container).toBeInTheDocument();

    const dots = document.querySelector('.Carousel__dots');
    expect(dots).toBeInTheDocument();
    expect(dots?.children.length).toEqual(2);

    const imgs = document.querySelector('.Carousel__img-wrapper');
    expect(imgs).toBeInTheDocument();
    expect(imgs?.children.length).toEqual(2);
  });

  it('If user passes Carousel component N inner elements following the correct syntax as below. If N=1, it should render a carousel that has 1 image and N dots', () => {
    const wrapper = render(
      <Carousel
        style={{
          width: 600,
          height: 240,
        }}
      >
        {imageSrc1.map((src, index) => (
          <div key={index}>
            <img
              src={src}
              style={{
                width: '100%',
              }}
            />
          </div>
        ))}
      </Carousel>
    );

    const container = document.querySelector('.Carousel');
    expect(container).toBeInTheDocument();

    const dots = document.querySelector('.Carousel__dots');
    expect(dots).toBeInTheDocument();
    expect(dots?.children.length).toEqual(1);

    const imgs = document.querySelector('.Carousel__img-wrapper');
    expect(imgs).toBeInTheDocument();
    expect(imgs?.children.length).toEqual(1);
  });
});
