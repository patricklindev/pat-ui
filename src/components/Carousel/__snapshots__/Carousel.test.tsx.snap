// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
    style="width: 600px; height: 240px;"
  >
    <div
      class="Carousel__img-wrapper"
    >
      <div
        id="Carousel__img__center"
        style="display: block;"
      >
        <img
          src="//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/cd7a1aaea8e1c5e3d26fe2591e561798.png~tplv-uwbnlip3yd-webp.webp"
          style="width: 100%;"
        />
      </div>
      <div
        id="Carousel__img__right"
        style="display: block;"
      >
        <img
          src="//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/6480dbc69be1b5de95010289787d64f1.png~tplv-uwbnlip3yd-webp.webp"
          style="width: 100%;"
        />
      </div>
      <div
        id="Carousel__img__left"
        style="display: block;"
      >
        <img
          src="//p1-arco.byteimg.com/tos-cn-i-uwbnlip3yd/24e0dd27418d2291b65db1b21aa62254.png~tplv-uwbnlip3yd-webp.webp"
          style="width: 100%;"
        />
      </div>
    </div>
    <span
      class="Carousel__prev"
    >
      ❮
    </span>
    <span
      class="Carousel__next"
    >
      ❯
    </span>
    <div
      class="Carousel__center Carousel__dots"
    >
      <span
        class="Carousel__dot Carousel__dot-active"
        id="0"
      />
      <span
        class="Carousel__dot"
        id="1"
      />
      <span
        class="Carousel__dot"
        id="2"
      />
      <span
        class="Carousel__dot"
        id="3"
      />
    </div>
  </div>
</DocumentFragment>
`;
